xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,250 @@
+{
+
+    private final Map<String, XWikiAttachment> map = new ConcurrentSkipListMap<String, XWikiAttachment>();
+
+    private final XWikiDocument document;
+
+    /**
+     * Initializes the map.
+     * 
+     * @since 10.0RC1
+     */
+    public XWikiAttachmentList(XWikiDocument document)
+    {
+        super(new ArrayList<XWikiAttachment>());
+        this.document = document;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        return set(attachment) != null;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index, XWikiAttachment attachment)
+    {
+        add(attachment);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public void clear()
+    {
+        updateMap();
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        updateMap();
+        return true;
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends XWikiAttachment> c)
+    {
+        return addAll(c);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.list.get(index).getFilename());
+        if (removedAttachment != null) {
+            removedAttachment = super.remove(index);
+            document.setMetaDataDirty(true);
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if(map.put(xwikiAttachment.getFilename(), xwikiAttachment) == null)
```

Comment:
I don't understand why you are calling put in a method supposed to remove something.

Reply:
@tmortagne you left a comment saying to call put in this method to check if the attachment exists in the list